File: Web/Nat/src/com/natera/util/DriverHelper.java
package com.natera.util;

import io.appium.java_client.AppiumDriver;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.openqa.selenium.By;
import org.openqa.selenium.ElementNotVisibleException;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.Select;
import org.testng.Assert;

public class DriverHelper
{

	protected WebDriver driver;
	protected AppiumDriver driver1;

	public DriverHelper(WebDriver webdriver)
	{
		driver = webdriver;	
		
	}

	public WebDriver getWebDriver()
	{
		return driver;
	}
	
	public AppiumDriver getMobileDriver()
	{
		driver1 = (AppiumDriver) driver;
		return driver1;
	}


	public void Log(String logMsg)
	{
		System.out.println(logMsg);
	}

	public By ByLocator(String locator) 
	{
		By result = null;

		if (locator.startsWith("//"))
		{
			result = By.xpath(locator);
		}
		else if (locator.startsWith("css=")) 
		{
			result = By.cssSelector(locator.replace("css=", ""));
		} 
		else if (locator.startsWith("#")) 
		{
			result = By.name(locator.replace("#", ""));			
		} 
		else if (locator.startsWith("link=")) {
			result = By.linkText(locator.replace("link=", ""));
		}
		else if(locator.startsWith("class=")) 
		{
			result=By.className(locator.replace("class=", ""));
		}
		else if(locator.startsWith("name=")) 
		{
			result=By.name(locator.replace("name=", ""));
		}		  
		else if(locator.startsWith("id=")) 
		{
			result=By.id(locator.replace("id=", ""));
		}
		else if(locator.startsWith("(")) {
			result = By.xpath(locator);
		}
		else 
		{
			result = By.id(locator);
		}
		return result;
	}


	public Boolean isElementPresent(String locator) 
	{
		Boolean result = false;
		try 
		{
			getWebDriver().findElement(ByLocator(locator));
			result = true;
		} catch (Exception ex) 
		{

		}

		return result;
	}


	public Boolean isElementDisplayed(String locator) 
	{

		Boolean result = false;
		try 
		{
			getWebDriver().findElement(ByLocator(locator)).isDisplayed();
			result = true;
		} catch (Exception ex) 
		{

		}
		return result;
	}


	public void WaitForElementPresent(String locator, int timeout)
	{

		for (int i = 0; i < timeout; i++) 
		{
			if (isElementPresent(locator)) 
			{
				break;
			}

			try 
			{
				Thread.sleep(1000);
			} catch (InterruptedException e) 
			{
				e.printStackTrace();
			}
		}
	}

	public void WaitForElementNotPresent(String locator, int timeout) 
	{

		for (int i = 0; i < timeout; i++) 
		{
			if (!isElementPresent(locator)) 
			{
				break;
			}

			try
			{
				Thread.sleep(1000);
			} catch (InterruptedException e)
			{
				e.printStackTrace();
			}
		}
	}
	public void WaitForElementEnabled(String locator, int timeout) 
	{
		for (int i = 0; i < timeout; i++) 
		{
			if (isElementPresent(locator)) 
			{
				if (getWebDriver().findElement(ByLocator(locator)).isEnabled()) 
				{
					break;
				}
			}

			try 
			{
				Thread.sleep(1000);
			} 
			catch (InterruptedException e)
			{
				e.printStackTrace();
			}
		}
	}

	public void WaitForElementNotEnabled(String locator, int timeout) 
	{

		for (int i = 0; i < timeout; i++) 
		{
			if (isElementPresent(locator)) 
			{
				if (!getWebDriver().findElement(ByLocator(locator)).isEnabled()) 
				{
					break;
				}
			}

			try 
			{
				Thread.sleep(1000);
			} catch (InterruptedException e) 
			{
				e.printStackTrace();
			}
		}
	}

	public void WaitForElementVisible(String locator, int timeout) 
	{

		for (int i = 0; i < timeout; i++) 
		{
			if (isElementPresent(locator)) 
			{
				if (getWebDriver().findElement(ByLocator(locator)).isDisplayed()) 
				{
					break;
				}
			}

			try 
			{
				Thread.sleep(1000);
			} catch (InterruptedException e) 
			{
				e.printStackTrace();
			}
		}
	}

	public void WaitForElementNotVisible(String locator, int timeout) 
	{

		for (int i = 0; i < timeout; i++) 
		{
			if (isElementPresent(locator)) 
			{
				if (!getWebDriver().findElement(ByLocator(locator)).isDisplayed()) 
				{
					break;
				}
			}

			try 
			{
				Thread.sleep(1000);
			} catch (InterruptedException e) 
			{
				e.printStackTrace();
			}
		}
	}

	public void mouseOver(String locator)
	{		
		this.WaitForElementPresent(locator, 50);		
		WebElement el = getWebDriver().findElement(ByLocator(locator));

		Actions builder = new Actions(getWebDriver());    
		builder.moveToElement(el).build().perform();		
	}

	public void mouseDoubleClick(String locator)
	{
		this.WaitForElementPresent(locator, 50);		
		WebElement el = getWebDriver().findElement(ByLocator(locator));
		Actions builder = new Actions(getWebDriver());    
		builder.doubleClick(el).perform();
	}

	public void dragAndDrop(String draggable, String to)
	{
		WebElement elDraggable = getWebDriver().findElement(ByLocator(draggable));
		WebElement todrag = getWebDriver().findElement(ByLocator(to));
		Actions builder = new Actions(getWebDriver());   
		builder.dragAndDrop(elDraggable, todrag).perform();		
	}

	public void clickOn(String locator)
	{		
		this.WaitForElementPresent(locator, 30);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		WebElement el = getWebDriver().findElement(ByLocator(locator));			
		el.click();
	}
	
	public void clickViaJavaScript(String locator)
	{
		WebElement element = driver.findElement(By.xpath(locator));
		JavascriptExecutor executor = (JavascriptExecutor)driver;
		executor.executeScript("arguments[0].click();", element);
	}

	public void clearTextField(String locator)
	{		
		this.WaitForElementPresent(locator, 30);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		WebElement el = getWebDriver().findElement(ByLocator(locator));			
		el.clear();
	}



	public void sendKeys(String locator, String userName)
	{		
		this.WaitForElementPresent(locator, 30);		
		Assert.assertTrue(isElementPresent(locator), "Element Locator :"+locator+" Not found");
		WebElement el = getWebDriver().findElement(ByLocator(locator));
		el.clear();
		el.sendKeys(userName);
	}


	public void selectDropDownByVisibleText(String locator, String targetValue)
	{ 
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		this.WaitForElementPresent(locator, 20);
		new Select(getWebDriver().findElement(ByLocator(locator))).selectByVisibleText(targetValue);		
	}

	public void selectDropDownByValue(String locator, String targetValue)
	{ 
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		this.WaitForElementPresent(locator, 20);
		new Select(getWebDriver().findElement(ByLocator(locator))).selectByValue(targetValue);		
	}

	public boolean isTextPresent(String locator, String str)
	{
		Assert.assertTrue(isElementPresent(locator) ,"Element Locator :"+locator+" Not found");

		String message = getWebDriver().findElement(ByLocator(locator)).getText();		
		System.out.println("my value--" + message);
		if(message.contains(str)){return true;}
		else {	return false; }
	}

	public String getText(String locator)
	{
		WaitForElementPresent(locator, 30);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		String text = getWebDriver().findElement(ByLocator(locator)).getText();			
		return text;
	}
	public void switchtoframe(String locator){
		driver.switchTo().frame(locator);
	}
	public boolean isChecked(String locator)
	{
		boolean checkStatus = false;
		WaitForElementPresent(locator, 20);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		WebElement el = getWebDriver().findElement(ByLocator(locator));	    
		checkStatus = el.isSelected();	    
		return checkStatus;

	}

	public String getAttribute(String locator, String attribute)
	{
		WaitForElementPresent(locator, 20);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		String text = getWebDriver().findElement(ByLocator(locator)).getAttribute(attribute);	
		return text;
	}
	public Integer getInt(String strString)
	{
		Pattern intsOnly = Pattern.compile("\\d+");
		String input = strString;
		Matcher makeMatch = intsOnly.matcher(input);
		makeMatch.find();
		String digitStr  = makeMatch.group();
		Integer digit = Integer.parseInt(digitStr);            
		return digit;
	}

	public void javaScriptExecute(String javascrpt)
	{
		JavascriptExecutor js = (JavascriptExecutor) getWebDriver();
		js.executeScript(javascrpt);
	}

	public void rightClick(String locator)
	{
		WebElement el = getWebDriver().findElement(ByLocator(locator));

		//build and perform the mouseOver with Advanced User Interactions API
		Actions builder = new Actions(getWebDriver());    
		builder.contextClick(el).build().perform();
	}


	//click on yes at the delete video popup
	public void acceptPopup()
	{
		getWebDriver().findElement(By.xpath("//button[contains(text(), 'Accept')]")).click();

	}

	public void clickTextField(String locator)
	{		
		this.WaitForElementPresent(locator, 30);
		Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
		WebElement el = getWebDriver().findElement(ByLocator(locator));			
		el.clear();
	}

	// scroll to Particular element
	public void scrollToElement(String locator) {
		WebElement element = getWebDriver().findElement(ByLocator(locator));
		JavascriptExecutor jse = (JavascriptExecutor)driver;
		jse.executeScript("arguments[0].scrollIntoView();", element);
	}

	public void scrollingdown()
	{
		JavascriptExecutor jse = (JavascriptExecutor)driver;
		jse.executeScript("window.scrollBy(0,250)", "");
		//((JavascriptExecutor) d).executeScript("window.scrollTo(0, document.body.scrollHeight)");
	}
	public void scrollingup()
	{
		JavascriptExecutor jse = (JavascriptExecutor)driver;
		jse.executeScript("window.scrollBy(0,0)", "");
	}

	public void waitForWorkArround(int Milisecond)
	{ 
		try 
		{
			Thread.sleep(Milisecond);
		}
		catch (InterruptedException e) 
		{
			e.printStackTrace();
		}
	}

	//**********************************************MOBILE METHODS**********************************************// 
	public void clickOnMobile(String locator)
	{		
		try
		{
			this.WaitForElementPresent(locator, 30);
			Assert.assertTrue(isElementPresent(locator),"Element Locator :"+locator+" Not found");
			WebElement el = getWebDriver().findElement(ByLocator(locator));			
			el.click();
		}
		catch(ElementNotVisibleException ex)
		{
			System.out.println("Running Test on System no need to use this button");
		}
	}
}

File: Web/Decryptex_Automation1/Decryptex_Automation/utils/DriverHelper.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Interactions;
using OpenQA.Selenium.Support.UI;
using System;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Linq;
using Decryptex_Automation.locators;
using System.Threading;

namespace Decryptex_Automation.Utils
{
	public class DriverHelper
	{
        public IWebDriver WebDriver { get; }

		public PropertyReader Properties { get; }

		public LocatorReader Locators { get; }

		public DriverHelper(IWebDriver driver, string locatorFile)
		{
			WebDriver = driver;
			Properties = new PropertyReader();

			if (locatorFile != null)
				Locators = new LocatorReader(locatorFile);
        }

		public By ByLocator(string locator)
		{
			By result;

			if (locator.StartsWith("//"))
			{
				result = By.XPath(locator);
			}
			else if (locator.StartsWith("css="))
			{
				result = By.CssSelector(locator.Replace("css=", ""));
			}
			else if (locator.StartsWith("#"))
			{
				result = By.Name(locator.Replace("#", ""));
			}
			else if (locator.StartsWith("Link="))
			{
				result = By.LinkText(locator.Replace("Link=", ""));
			}
			else if (locator.StartsWith("class="))
			{
				result = By.ClassName(locator.Replace("class=", ""));
			}
			else if (locator.StartsWith("name="))
			{
				result = By.Name(locator.Replace("name=", ""));
			}
			else if (locator.StartsWith("id="))
			{
				result = By.Id(locator.Replace("id=", ""));
			}
			else if (locator.StartsWith("("))
			{
				result = By.XPath(locator);
			}
			else
			{
				result = By.XPath(locator);
			}
			return result;
		}

		public object ExecuteScript(string javascrpt, params object[] args)
		{
			IJavaScriptExecutor js = (IJavaScriptExecutor)WebDriver;
			return js.ExecuteScript(javascrpt, args);
		}

		public TResult Wait<TResult>(Func<IWebDriver, TResult> predicate, int timeoutSeconds)
		{
			WebDriverWait wait = new WebDriverWait(WebDriver, TimeSpan.FromSeconds(timeoutSeconds));
            wait.IgnoreExceptionTypes(typeof(StaleElementReferenceException));
            wait.IgnoreExceptionTypes(typeof(InvalidElementStateException));
            wait.PollingInterval = TimeSpan.FromMilliseconds(timeoutSeconds);
			return wait.Until(predicate);
		}

		// find an element satisfying a specific predicate allowing for some time before the condition is satisfied
		public IWebElement FindElement(string locator, Func<By, Func<IWebDriver, IWebElement>> predicate, int timeoutSeconds, bool throwError = true)
		{
			try
			{
				return Wait(predicate(ByLocator(locator)), timeoutSeconds);
			}
			catch
			{
				if (throwError)
					throw;
			}

			return null;
		}

		// find all elements satisfying a specific predicate allowing for some time before the condition is satisfied
		public ReadOnlyCollection<IWebElement> FindElements(string locator, Func<By, Func<IWebDriver, ReadOnlyCollection<IWebElement>>> predicate, int timeoutSeconds, bool throwError = true)
		{
			try
			{
				WebDriverWait wait = new WebDriverWait(WebDriver, TimeSpan.FromSeconds(timeoutSeconds));
				return wait.Until(predicate(ByLocator(locator)));
			}
			catch
			{
				if (throwError)
					throw;
			}

			return null;
		}

        //Get available elements count
		public int GetElementCount(string locator)
		{
			ReadOnlyCollection<IWebElement> elements = FindElements(locator, ExpectedConditions.PresenceOfAllElementsLocatedBy, 10, false);
			return elements == null ? 0 : elements.Count;
		}

        //Verify element is present or not
		public bool IsElementPresent(string locator)
		{
			return FindElement(locator, ExpectedConditions.ElementExists, 3, false) != null;
		}

        //Wait until element doesn't hide
        public void WaitForElementHide(string locator,int timeoutSeconds=10)
        {
            GetWait(timeoutSeconds).Until(ExpectedConditions.InvisibilityOfElementLocated(ByLocator(locator)));
        }

        //Wait until elemenent doesn't display
        public void WaitForElementDisplay(string locator, int timeoutSeconds=10)
        {

            GetWait(timeoutSeconds).Until(ExpectedConditions.ElementIsVisible(ByLocator(locator)));
        }

        //Wait for element element until it gets enable
        public void waitForElementEnabled(string locator,int timeoutSeconds=10)
        {
            IWebElement el = FindElement(locator, ExpectedConditions.ElementToBeClickable, timeoutSeconds);

            GetWait(timeoutSeconds).Until(drv => el.Enabled);
        }

        //Verify element is enabled or not
        public bool IsElementEnabled(string locator)
		{
			return FindElement(locator, ExpectedConditions.ElementToBeClickable, 3, false) != null;
		}

        //Verify element is visible or not
		public bool IsElementDisplayed(string locator,int timeout=3)
		{
			return FindElement(locator, ExpectedConditions.ElementIsVisible, timeout, false) != null;
		}

        //Perform mouse over action on an element
		public void MouseOver(string locator)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 10);

			Actions builder = new Actions(WebDriver);
			builder.MoveToElement(el).Build().Perform();
		}

        //Perform drag and drop action
		public void DragAndDrop(string draggable, string to)
		{
			IWebElement elDraggable = FindElement(draggable, ExpectedConditions.ElementIsVisible, 10);
			IWebElement todrag = FindElement(to, ExpectedConditions.ElementIsVisible, 10);

			Actions builder = new Actions(WebDriver);
			builder.DragAndDrop(elDraggable, todrag).Perform();
		}

        //Simple click on an element
		public void Click(string locator, params string[] modifiers)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementToBeClickable, 10);
            el = FindElement(locator, ExpectedConditions.ElementIsVisible, 10);

            Actions action = new Actions(WebDriver);
			// press modifier keys (if any)
			if (modifiers != null && modifiers.Length > 0)
			{
				foreach (string key in modifiers)
					action = action.KeyDown(key);
			}

			el.Click();

			// release modifier keys (if any)
			if (modifiers != null && modifiers.Length > 0)
			{
				foreach (string key in modifiers.Reverse())
					action = action.KeyUp(key);
			}
		}
        
        //Click on an element using javaScript
		public void ClickByJS(string locator)
		{
            IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 10); 
            ExecuteScript("arguments[0].click();", el);
        }

        //Select value from dropdown
		public void SelectDropDown(string locator, string targetValue)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementToBeClickable, 10);
			new SelectElement(el).SelectByText(targetValue);
			Wait(d => ExpectedConditions.ElementToBeSelected(el), 5);
		}

        //Verify element contains text or not
		public bool IsTextPresent(string locator, string str)
		{
			return GetText(locator).Contains(str);
		}

        //Get available text of an element
        //Updated time from 5 to 15 because element was not being visible in 5 sec.
        //While using ElementExists we were getting blank as the test so used ElementIsVisible
        public string GetText(string locator)
		{
            //IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 10);
            IWebElement el = FindElement(locator, ExpectedConditions.ElementIsVisible, 10);
            return el.Text;
		}

        //Enter data in the field without clearing it
		public void SendKeys(string locator, string text, bool clear = false)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementToBeClickable, 15);

			if (clear)
				el.Clear();

			el.SendKeys(text);
		}

        //Enter data in the field after clearing it
		public void SetText(string locator, string text)
		{
			SendKeys(locator, text, true);
		}

        //Some time selenium sendkeys was giving error so used this method to set text field value 
		public void SetTextByJS(string locator, string text)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 15);
			ExecuteScript("arguments[0].setAttribute('value', '" + text + "')", el);
			el.SendKeys(Keys.Enter);
		}

        //Verify element checked or not
		public bool IsChecked(string locator)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementIsVisible, 15);
			return el.Selected;
		}

        //Get value of an attribute of given element
		public string GetAttribute(string locator, string attribute)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 15);
			return el.GetAttribute(attribute);
		}

        //Wait untill full page load
		public void WaitForPageLoad(string title, int timeoutSeconds = 10)
		{
            //Updated : added javascript code to wait untill full document loaded
            Wait(ExpectedConditions.TitleContains(title), timeoutSeconds);
            Wait(d => ((IJavaScriptExecutor)WebDriver).ExecuteScript("return document.readyState").Equals("complete"), timeoutSeconds);
        }

        //In small screen/ low resolution full grid doesn't displayed this method scroll the age to the down
		public void ScrollToBottom()
		{
			ExecuteScript("window.scrollTo(0, document.body.scrollHeight);");
		}

        //This method scroll the age to the top
        public void ScrollToTop()
		{
            //ExecuteScript("window.scrollTo(0,0)");
            ExecuteScript("window.scrollBy(0,-300)");

        }

		public void ScrollIntoView(string locator)
		{
			IWebElement el = FindElement(locator, ExpectedConditions.ElementExists, 15);
			ExecuteScript("arguments[0].scrollIntoView();", el);
		}

        //wait for frame to be available and switch into it
		public void SwitchToFrame(string xpath)
		{
			Wait(ExpectedConditions.FrameToBeAvailableAndSwitchToIt(xpath), 15);
		}

        //To get back from frame
		public void DefaultSwitch()
		{
			WebDriver.SwitchTo().DefaultContent();
		}

        //Get current project path
		public string GetPath()
		{
			return Properties.GetPath();
		}

        //Update System time zone
        public void SetSystemTimeZone(string timeZoneId)
        {
            var process = Process.Start(new ProcessStartInfo
            {
                FileName = "tzutil.exe",
                Arguments = "/s \"" + timeZoneId + "\"",
                UseShellExecute = false,
                CreateNoWindow = true
            });

            if (process != null)
            {
                process.WaitForExit();
                TimeZoneInfo.ClearCachedData();
            }
        }

        public WebDriverWait GetWait(int timeOutSeconds)
        {
            return new WebDriverWait(WebDriver, TimeSpan.FromSeconds(timeOutSeconds));
        }

        //Perform Zoom in-out action
        public void Zoom(int percentage = 100)
        {
            ExecuteScript("document.body.style.zoom = '" + percentage + "%';");
        }
    }
}